From d337a44e429e6de23ed3d73fcb81ec44f7b05522 Mon Sep 17 00:00:00 2001
From: Marios Makassikis <mmakassikis@freebox.fr>
Date: Tue, 27 Jul 2021 09:24:51 +0900
Subject: [PATCH] ksmbd: Return STATUS_OBJECT_PATH_NOT_FOUND if smb2_creat()
 returns ENOENT
Git-commit: d337a44e429e6de23ed3d73fcb81ec44f7b05522
Patch-mainline: v5.15-rc1
References: bsn#12
Modified-by-SEL: Yes, modified due to different context


Both Windows 10's SMB server and samba return STATUS_OBJECT_PATH_NOT_FOUND
when trying to access a nonexistent path.

This fixes Windows 10 File History tool. The latter relies on the server
returning STATUS_OBJECT_PATH_NOT_FOUND to figure out what part of the
target path needs to be created. Returning STATUS_OBJECT_NAME_INVALID
will make it stop and display an error to the user.

Signed-off-by: Marios Makassikis <mmakassikis@freebox.fr>
Signed-off-by: Namjae Jeon <namjae.jeon@samsung.com>
Signed-off-by: Steve French <stfrench@microsoft.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/ksmbd/smb2pdu.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- a/fs/ksmbd/smb2pdu.c
+++ b/fs/ksmbd/smb2pdu.c
@@ -2755,8 +2755,13 @@ int smb2_open(struct ksmbd_work *work)
 	if (!file_present) {
 		rc = smb2_creat(work, &path, name, open_flags, posix_mode,
 				req->CreateOptions & FILE_DIRECTORY_FILE_LE);
-		if (rc)
+		if (rc) {
+			if (rc == -ENOENT) {
+				rc = -EIO;
+				rsp->hdr.Status = STATUS_OBJECT_PATH_NOT_FOUND;
+			}
 			goto err_out;
+		}
 
 		created = true;
 		if (ea_buf) {
